File: manpassd/passcrypto/passcrypto.go
/*
This package provide crypto services, most high level entry functions are:
- EncryptMe: use scrypt to generate encrypt key for each encryption, and append salt to the result cipher; input clear text is a slice of byte
- DecryptMe
- EncryptWithoutSalt: doesn't use scrypt, expect user to re-use the scrypt generated key;much faster but less secure
- DecryptWithoutSalt

The underlying crypto is nacl/secretbox
*/
package passcrypto

import (
	"crypto/rand"
	"encoding/base32"
	"errors"
	"golang.org/x/crypto/nacl/secretbox"
	"golang.org/x/crypto/scrypt"
	"io"
)

var (
	ErrEncrypt = errors.New("secret: encryption failed")
	ErrDecrypt = errors.New("secret: decryption failed")
)

const (
	KeySize   = 32
	NonceSize = 24
	SaltSize  = 16
	Scrypt_N  = 1024 //1048576 is for file encryption, 16384 is for interactive logins
	Scrypt_r  = 1
	Scrypt_p  = 1
)

func ZeroMe(myr []byte) {
	//zero a byte slice
	for k, _ := range myr {
		myr[k] = 0
	}
	return
}

func GenerateNonce() (*[NonceSize]byte, error) {
	nonce := new([NonceSize]byte)
	_, err := io.ReadFull(rand.Reader, nonce[:])
	if err != nil {
		return nil, err
	}
	return nonce, nil
}

func GenerateSalt() (*[SaltSize]byte, error) {
	salt := new([SaltSize]byte)
	_, err := io.ReadFull(rand.Reader, salt[:])
	if err != nil {
		return nil, err
	}
	return salt, nil
}

func GenerateEncKey(passwd []byte, salt *[SaltSize]byte) (*[KeySize]byte, error) {
	var rkey = new([KeySize]byte)
	skey, err := scrypt.Key(passwd, salt[:], Scrypt_N, Scrypt_r, Scrypt_p, KeySize)
	if err != nil {

		return nil, err
	}
	copy(rkey[:], skey)
	ZeroMe(skey)
	return rkey, nil
}

func EncryptWithoutSalt(msg []byte, skey *[KeySize]byte) ([]byte, error) {
	nonce, err := GenerateNonce()
	if err != nil {
		return nil, ErrEncrypt
	}
	out := make([]byte, len(msg)+NonceSize)
	out = secretbox.Seal(nonce[:], []byte(msg), nonce, skey)
	return out, nil

}

func EncryptMe(msg []byte, inpasswd []byte) ([]byte, error) {
	salt, err := GenerateSalt()
	if err != nil {
		return nil, ErrEncrypt
	}
	skey, err := GenerateEncKey(inpasswd, salt)
	if err != nil {
		return nil, ErrEncrypt
	}
	rstr, err := EncryptWithoutSalt(msg, skey)
	if err != nil {
		return nil, ErrEncrypt
	}
	return append(salt[:], rstr...), nil
}

func EncryptMeBase32(msg []byte, inpasswd []byte) (string, error) {
	cipher, err := EncryptMe(msg, inpasswd)
	if err != nil {
		return "", err
	}
	return base32.StdEncoding.EncodeToString(cipher), nil
}

func DecryptWithoutSalt(cipher []byte, skey *[KeySize]byte) ([]byte, error) {
	var nonce = [NonceSize]byte{}
	copy(nonce[:], cipher[:NonceSize])
	out, ok := secretbox.Open(nil, cipher[NonceSize:], &nonce, skey)
	if !ok {
		return nil, ErrDecrypt
	}
	return out, nil
}

func DecryptMe(cipher []byte, inpasswd []byte) ([]byte, error) {
	var salt = [SaltSize]byte{}
	copy(salt[:], cipher[:SaltSize])
	skey, err := GenerateEncKey(inpasswd, &salt)
	if err != nil {
		return nil, ErrDecrypt
	}
	ostr, err := DecryptWithoutSalt(cipher[SaltSize:], skey)
	if err != nil {
		return nil, ErrDecrypt
	}
	return ostr, nil
}

func DecryptMeBase32(cipher_b32 string, inpasswd []byte) ([]byte, error) {
	cipher, err := base32.StdEncoding.DecodeString(cipher_b32)
	if err != nil {
		return nil, err
	}
	return DecryptMe(cipher, inpasswd)

}

File: manpassc/passcrypto.py
import pyscrypt
import nacl
import nacl.secret
import nacl.utils
import nacl.encoding
import nacl.hash

KeySize   = 32
NonceSize = 24
SaltSize  = 16
Scrypt_N  = 1024
Scrypt_r  = 1
Scrypt_p  = 1

def GenerateSalt():
    return nacl.utils.random(SaltSize)

def GenerateEncKey(passwd, salt):
    enc_key=pyscrypt.hash(password=passwd,
        salt = salt,
        N=Scrypt_N,
        r=Scrypt_r,
        p=Scrypt_p,
        dkLen=KeySize)
    return enc_key

def GenerateNonce():
    return nacl.utils.random(NonceSize)


def EncryptWithoutSalt(msg,skey):
    nonce=GenerateNonce()
    box = nacl.secret.SecretBox(skey)
    cipher_txt=box.encrypt(msg,nonce)
    return cipher_txt

def DecryptWithoutSalt(cipher,skey):
    box = nacl.secret.SecretBox(skey)
    return box.decrypt(cipher)

def EncryptMe(msg, passwd):
    salt=GenerateSalt()
    skey=GenerateEncKey(passwd,salt)
    rstr=EncryptWithoutSalt(msg,skey)
    return salt+rstr

def DecryptMe(cipher,passwd):
    salt=cipher[:SaltSize]
    skey=GenerateEncKey(passwd,salt)
    return DecryptWithoutSalt(cipher[SaltSize:],skey)

def EncryptMeBase32(msg,passwd):
    cipher=EncryptMe(msg,passwd)
    enc=nacl.encoding.Base32Encoder()
    return enc.encode(cipher)

def EncryptWithoutSaltBase32(msg,skey,salt):
    cipher=EncryptWithoutSalt(msg,skey)
    enc=nacl.encoding.Base32Encoder()
    return enc.encode(salt+cipher)

def DecryptMeBase32(cipher,passwd):
    enc=nacl.encoding.Base32Encoder()
    return DecryptMe(enc.decode(cipher),passwd)



def HashMsg(msg,skey):
    h=nacl.hash.sha256(skey+msg)
    return h+msg

def VerifyHash(msg,skey):
    if len(msg)<=64:
        return False
    h=msg[0:64]
    nh=nacl.hash.sha256(skey+msg[64:])
    if nh==h:
        return msg[64:]
    else:
        return False



if __name__ == '__main__':
    s=HashMsg("xixixi","z123")
    print VerifyHash(s+"1","z123")
##    clear="hello world!"
##    passwd="alu123"
##    s=EncryptMeBase32(clear,passwd)
##    if DecryptMeBase32(s,passwd) != clear:
##        print "error!"
##    else:
##        print "ok"
##    inf=open("d:\\temp\\tls\\root.cer","rb")
##    clear=inf.read()
##    inf.close()
##    outf=open("d:\\temp\\tls\\root.encrypted","w")
##    outf.write(EncryptMeBase32(clear,"alu123"))
##    outf.close()
